packages/@lwc/engine-dom/src/renderer.ts

```diff
@@ -20,6 +20,20 @@ const globalStylesheetsParentElement: Element = document.head || document.body |
 
 let getCustomElement, defineCustomElement, HTMLElementConstructor;
 
+// Read the existing styles from the document
+// This could be enhanced if we associate a hash to every style tag. A unique hash can be generated by the compiler
+// and passed to insertGlobalStylesheet, then set as the data-lwc-ssr value. This way, we would only compare simple attribute
+// values and not the style content.
+(function readExistingStyles() {
+    const styles = globalStylesheetsParentElement.querySelectorAll('style[data-lwc-ssr]');
```

Comment:
This is in my opinion the most contentious part of this PR. 

After thinking about this, we shouldn't make a distinction between stylesheets inject by the server and the one injected at runtime. The `insertGlobalStylesheet` is invoked at runtime for synthetic shadow stylesheet. IMO, the styles injected by `insertGlobalStylesheet` should have the same attribute as the one injected by SSR. What about renaming this attribute to `lwc-synthetic-shadow-stylesheet` or something like that?

How the components are rehydrated is orthogonal to this PR, I would be in favor of extracting this except if we are able to reach an agreement here. =)

Reply:
I renamed the attribute `lwc-scoped-style`. I agree that the client runtime should use it as well, to handle the case where the same components are added twice from different bundle.
Note sure what the contentious issue is. Please let me know what you want to do.